### ISSUE TITLE
[QUESTION] How do I run a mass conservation test?

### ISSUE THREAD

**Issue by LiamBindle** [geoschem/GCHP#40]
Hi everyone,

In the GCSC meeting the other day, it was suggested I test mass conservation in stretched-grid simulations. Could someone help me understand how I could do this? I see there's a `geosfp_2x25_masscons` run directory for GC-Classic&mdash;is there a way to mirror this in GCHPctm?

Thanks in advance!

**Comment by lizziel** [comment]
The old GEOS-Chem Classic mass conservation run created a text file with global mass of a passive tracer every 6 hrs. The code was GC-Classic-specific since it did not sum across multiple nodes, and was retired early this year. What you can do instead is increase the checkpoint file output frequency and calculate global mass from that. The transport tracer simulation checkpoint file should have the meteorology data needed to convert mol/mol to kg.

**Comment by sdeastham** [comment]
It should also be possible to perform an `MPI_Reduce` call within GCHP, and then having the root CPU report the result. That would be less output intensive.

**Comment by lizziel** [comment]
If you go the `MPI_Reduce` route I suggest adding it to the `print_global_species_kg` subroutine in `GeosUtil/unitconv_mod.F90`. That subroutine exists to do quick n' dirty mass conservation validation online but was never extended to work with GCHP.

Regardless, we should include mass conservation as part of our benchmarking process. With the removal of the `masscons` option in 12.8.0 we no longer do that. I created a feature request on gcpy to bring in a replacement for it done as part of benchmarking post-processing: https://github.com/geoschem/gcpy/issues/93.

**Comment by LiamBindle** [comment]
> It should also be possible to perform an MPI_Reduce call within GCHP, and then having the root CPU report the result. That would be less output intensive.

Can we get a 1D timeseries diagnostic from HISTORY? By that I mean a netcdf file with a multiple times in it? If so, that seems like a clean solution, and I'd be interested in implementing that.

In the meantime (if 1D timeseries are possible), post-processing checkpoint files sounds like a good practical way to get this done.

**Comment by lizziel** [comment]
That's also definitely a way to go and theoretically should work. It's not a solution for GC-Classic, however, which this issue also applies to. But that shouldn't stop trying to do it.

**Comment by LiamBindle** [comment]
Unfortunatley I don't have the time to implement a longterm solution for this, but the following is a quick and effective way to run a mass conservation test in GCHP.

Add the following collection
```
  MassCons.template:      '%y4%m2%d2_%h2%n2z.nc4',
  MassCons.format:        'CFIO',
  MassCons.frequency:     240000
  MassCons.duration:      240000
  MassCons.mode:          'instantaneous'
  MassCons.fields:        'Met_AD', 'GCHPchem',
                          'SpeciesConc_PassiveTracer', 'GCHPchem',
::
```

Run the simulation for 1-year. After 1-year run the following NCOs in `OutputDir/`.

```console
$ ncrcat GCHP.MassCons.* masscons.nc
$ ncap2 --cnk_dmn time,100 -s 'MolesPassiveTracer=Met_AD/0.02897*SpeciesConc_PassiveTracer; TotalMolesPassiveTracer=MolesPassiveTracer.total($lev,$nf,$Ydim,$Xdim); TotalDryMass=Met_AD.total($lev,$nf,$Ydim,$Xdim)' masscons.nc masscons-post.nc
$ ncks -v 'TotalMolesPassiveTracer,TotalDryMass' masscons-post.nc totalmass.nc
```

This will create a 1D timeseries of total dry mass and total moles of the passive tracer. 

### Results

#### Passive Tracer
My test was specifically looking at whether there mass conservation of stretched-grid simulations was different from cubed-sphere simulations. The following is the 1D timeseries for total moles of the passive tracer.

![image](https://user-images.githubusercontent.com/12772306/96150455-fcf70b80-0ecf-11eb-90e0-7653b2c83b96.png)

Note that the y-axis is highly zoomed in. Relative differences between C90, C48, and S48 (C48 w/ SF=2) are on the order of 1e-7. In my interpretation, deviations seem on the order of what would be expected.

The relative change from the initial total moles was  
![image](https://user-images.githubusercontent.com/12772306/96150633-2f086d80-0ed0-11eb-9b25-24c04ce7959d.png)

#### Atmosphere's total dry mass

It appears most changes come from differences in total dry mass. The 1D timeseries for total dry mass is
![image](https://user-images.githubusercontent.com/12772306/96150802-5cedb200-0ed0-11eb-9156-296572cc9cea.png)

The relative differences in total dry mass are more significant (up to 0.5% between C48 and C90). Deviations from the initial mass were on the order of 1e-5.
![image](https://user-images.githubusercontent.com/12772306/96151309-e43b2580-0ed0-11eb-9b91-a3f039252168.png)


In summary, there doesn't appear to be any sigifncant differences in mass conservation for stretched-grid simulations.
